src/data_conn.c

```diff
@@ -0,0 +1,612 @@
+				r->ongoing_io_len = 0;
+				r->ongoing_io_buf = NULL;
+			}
+			else {
+				r->ongoing_io_len = resp_hdr->len;
+				r->ongoing_io_buf = malloc(resp_hdr->len);
+			}
+			/* Fall through */
+		case READ_IO_RESP_DATA:
+			reqlen = r->ongoing_io_len - (r->io_read);
+			resp_data = r->ongoing_io_buf;
+			if (reqlen != 0) {
+				count = perform_read_on_fd(fd, ((uint8_t *)(resp_data)) + (r->io_read),
+				    reqlen);
+
+				if (count == -1)
+					return -1;
+				r->io_read += count;
+				if (count != (ssize_t)reqlen)
+					return READ_PARTIAL;
+			}
+			return READ_COMPLETED;
+	}
+	return -1;
+}
+
+int
+write_cmd(int fd, rcmd_t *cmd)
+{
+	ssize_t rc;
+	int err, i;
+
+start:
+	rc = writev(fd, cmd->iov, cmd->iovcnt);
+	err = errno;
+	if (rc < 0) {
+		if (err == EINTR)
+			goto start;
+		if ((err != EAGAIN) && (err != EWOULDBLOCK))
+			return -1;
+		return WRITE_PARTIAL;
+	}
+
+	for (i = 0; i < cmd->iovcnt; i++) {
+		if (cmd->iov[i].iov_len != 0 && cmd->iov[i].iov_len > (size_t)rc) {
+			cmd->iov[i].iov_base = (void *)(((uint8_t *)cmd->iov[i].iov_base) + rc);
+			cmd->iov[i].iov_len -= rc;
+			break;
+		}
+		else {
+			rc -= cmd->iov[i].iov_len;
+			cmd->iov[i].iov_len = 0;
+		}
+	}
+
+	if (i == cmd->iovcnt)
+		return WRITE_COMPLETED;
+
+	/* This is the case where
+	 * - writev wrote less
+	 * - there is more data to write
+	 * - and writev can write more */
+	goto start;
+}
+
+int
+handle_data_eventfd(void *arg)
+{
+	replica_t *r = (replica_t *)arg;
+	rcmd_t *cmd;
+	int ret;
+
+	while ((cmd = dequeue_replica_cmdq(r)) != NULL)
+		move_to_blocked_or_ready_q(r, cmd);
+	ret = handle_epoll_out_event(r);
+	return ret;
+}
+
+static int
+handle_mgmt_eventfd(void *arg)
+{
+	replica_t *r = (replica_t *) arg;
+	if (r->disconnect_conn > 0)
+		return -1;
+	return 0;
+}
+
+int
+do_drainfd(int data_eventfd)
+{
+	uint64_t value;
+	int rc, err;
+	while (1) {
+		rc = read(data_eventfd, &value, sizeof (value));
+		err = errno;
+		if (rc < 0) {
+			if (err == EINTR)
+				continue;
+			if (err != EAGAIN && err != EWOULDBLOCK)
+				return -1;
+			break;
+		}
+		if (rc == 0) {
+			return -1;
+		}
+	}
+	return 0;
+}
+
+int
+handle_epoll_in_event(replica_t *r)
+{
+	int ret, idx;
+	rcommon_cmd_t *rcomm_cmd;
+	bool task_completed = false;
+	bool unblocked = false;
+	pthread_mutex_t *mtx;
+	pthread_cond_t *cond_var;
+
+start:
+	ret = read_cmd(r);
+	if (ret < 0)
+		return -1;
+
+	if (ret == READ_COMPLETED) {
+		idx = r->ongoing_io->idx;
+		TAILQ_REMOVE(&r->waitq, r->ongoing_io, next);
+		rcomm_cmd = r->ongoing_io->rcommq_ptr;
+
+		mtx = rcomm_cmd->mutex;
+		cond_var = rcomm_cmd->cond_var;
+
+		rcomm_cmd->resp_list[idx].io_resp_hdr = *(r->io_resp_hdr);
+		rcomm_cmd->resp_list[idx].data_ptr = r->ongoing_io_buf;
+		if (rcomm_cmd->opcode == ZVOL_OPCODE_WRITE)
+			rcomm_cmd->resp_list[idx].data_len = rcomm_cmd->data_len;
+		else
+			rcomm_cmd->resp_list[idx].data_len = r->ongoing_io_len;
+		rcomm_cmd->resp_list[idx].status = 1;
+
+		MTX_LOCK(mtx);
```

Comment:
as we already taking the lock, lets move the lock few statements above to update resp_list also
also, add comment to do atomic fetch and updates in case of removing the locks

Reply:
same with error case also